src/top_down_iterator.jl

```diff
@@ -129,280 +128,102 @@ Currently, there are two possible causes of the expansion failing:
 """
 @enum ExpandFailureReason limit_reached=1 already_complete=2
 
-
-"""
-    @enum PropagateResult tree_complete=1 tree_incomplete=2 tree_infeasible=3
-
-Representation of the possible results of a constraint propagation. 
-At the moment there are three possible outcomes:
-
-- `tree_complete`: The propagation was applied successfully and the tree does not contain any holes anymore. Thus no constraints can be applied anymore.
-- `tree_incomplete`: The propagation was applied successfully and the tree does contain more holes. Thus more constraints may be applied to further prune the respective domains.
-- `tree_infeasible`: The propagation was succesful, but there are holes with empty domains. Hence, the tree is now infeasible.
-"""
-@enum PropagateResult tree_complete=1 tree_incomplete=2 tree_infeasible=3
-
-TreeConstraints = Tuple{AbstractRuleNode, Set{LocalConstraint}, PropagateResult}
-IsValidTree = Bool
-
-"""
-    struct PriorityQueueItem 
-
-Represents an item in the priority enumerator priority queue.
-An item contains of:
-
-- `tree`: A partial AST
-- `size`: The size of the tree. This is a cached value which prevents
-   having to traverse the entire tree each time the size is needed.
-- `constraints`: The local constraints that apply to this tree. 
-   These constraints are enforced each time the tree is modified.
-"""
-struct PriorityQueueItem 
-    tree::AbstractRuleNode
-    size::Int
-    constraints::Set{LocalConstraint}
-    complete::Bool
-end
-
-"""
-    PriorityQueueItem(tree::AbstractRuleNode, size::Int)
-
-Constructs [`PriorityQueueItem`](@ref) given only a tree and the size, but no constraints.
-"""
-PriorityQueueItem(tree::AbstractRuleNode, size::Int) = PriorityQueueItem(tree, size, [])
-
-
 """
     Base.iterate(iter::TopDownIterator)
 
 Describes the iteration for a given [`TopDownIterator`](@ref) over the grammar. The iteration constructs a [`PriorityQueue`](@ref) first and then prunes it propagating the active constraints. Recursively returns the result for the priority queue.
 """
 function Base.iterate(iter::TopDownIterator)
     # Priority queue with number of nodes in the program
-    pq :: PriorityQueue{PriorityQueueItem, Union{Real, Tuple{Vararg{Real}}}} = PriorityQueue()
+    pq :: PriorityQueue{State, Union{Real, Tuple{Vararg{Real}}}} = PriorityQueue()
 
-    grammar, max_depth, max_size, sym = iter.grammar, iter.max_depth, iter.max_size, iter.sym
+    #TODO: instantiating the solver should be in the program iterator macro
+    if isnothing(iter.solver)
+        iter.solver = Solver(iter.grammar, iter.sym)
```

Comment:
This is totally fine. However, we should rather check that the solver is instantiated during instantiation of the iterator.